migrations/1585388899167-add_default_file_storage_provider.js

```diff
@@ -0,0 +1,76 @@
+const mongoose = require('mongoose');
+const CryptoJS = require('crypto-js');
+const { Configuration } = require('@schul-cloud/commons');
+
+const { info, error } = require('../src/logger');
+const { schoolModel: School } = require('../src/services/school/model');
+
+const { connect, close } = require('../src/utils/database');
+
+const StorageProvider = mongoose.model('storageProviderMigration', new mongoose.Schema({
+	type: { type: String, enum: ['S3'], required: true },
+	isShared: { type: Boolean },
+	accessKeyId: { type: String, required: true },
+	secretAccessKey: { type: String, required: true },
+	endpointUrl: { type: String, required: true },
+	region: { type: String },
+	maxBuckets: { type: Number, required: true },
+	schools: [{ type: mongoose.Schema.Types.ObjectId, ref: 'school' }],
+}, {
+	timestamps: true,
+}), 'storageproviders');
+
+module.exports = {
+	up: async function up() {
+		await connect();
+		const S3_KEY = Configuration.get('S3_KEY');
+
+		if (!process.env.AWS_SECRET_ACCESS_KEY) {
+			info('No AWS config found. Migration will be successful but won\'t change anything!');
+		} else {
+			if (!S3_KEY) {
+				error('You need to set process.env.S3_KEY to encrypt the old key!');
+			}
+
+			let schools = await School.find({})
+				.select(['_id'])
+				.lean({ virtuals: true })
+				.exec();
+			schools = schools.map((s) => s._id);
+			info(`Got ${schools.length} schools.`);
+
+			const providers = await StorageProvider.find({ $where: 'this.schools.length > 0' })
+				.select(['schools'])
+				.lean()
+				.exec();
+
+			const assignedSchools = [...providers.map((p) => p.schools)].map((s) => s.toString());
+			info(`Got ${assignedSchools.length} schools that already use another provider.`);
+
+			const secretAccessKey = CryptoJS.AES.encrypt(process.env.AWS_SECRET_ACCESS_KEY, S3_KEY);
+			const provider = new StorageProvider({
+				type: 'S3',
+				isShared: true,
+				accessKeyId: process.env.AWS_ACCESS_KEY,
```

Comment:
`endpointUrl , and accessKeyId` should be indexed and not occure twice.
imagine the migration would be executed a secont time, it should reuse the existing provider...

Reply:
Ich glaube in Production können wir sicher sein, dass die Migration nicht zweimal ausgeführt wird. Unique müssen die glaube ich nicht unbedingt sein (Usecase Minio, bzw. überlappender AccessKey bei unterschiedlichen Providern)